accepted/2019/support-for-memory-limits.md

```diff
@@ -1,58 +1,122 @@
+It is critical to provide effective and well-defined capabilities for .NET applications within memory-limited cgroups. An application should run indefinitely given a sensible configuration for that application. It is important that .NET developers have good controls to optimize their container hosted applications. Our goal is that certain classes of .NET applications can be run with <100 MB memory constraints.
 
-See [implementing hard limit for GC heap dotnet/coreclr #22180](https://github.com/dotnet/coreclr/pull/22180).
+Related:
+
+- [implementing hard limit for GC heap dotnet/coreclr #22180](https://github.com/dotnet/coreclr/pull/22180).
+- [Validate container improvements with .NET 6](https://github.com/dotnet/runtime/issues/53149).
+
+## Cgroup constraints
+
+cgroups control two main resources: memory and cores. Both are relevant to the .NET GC.
+
+Memory constraints defines the maximum memory available to the cgroup. This memory is used by the guest operating system, the .NET runtime, the GC heap, and potentially other users. If a cgroup has `100 MB` available, the app will have less than that. The cgroup will be terminated (AKA `OOMKilled`) when the memory limit is reached.
+
+Core constraints determine how many GC heaps should be created, at maximum. The maximum heap value matches `Environment.ProcessorCount`. There are three primary ways that this value can be set (using the `docker` CLI to demonstrate):
+
+- Not specified -- `Environment.ProcessorCount` will match the total number of machine cores.
+- Via `--cpus` -- `Environment.ProcessorCount` uses that (decimal) value (rounded up to the next integer).
+- Via `--cpu-sets` -- `Environment.ProcessorCount` matches the count of specified CPUs.
+- `DOTNET_PROCESSOR_COUNT` -- `Environment.ProcessorCount` uses this value. If other values are also specified, they are ignored.
+
+In the general case, there will be one heap per core. If the GC creates too many heaps, that can over-eagerly use up the memory limit, at least in part. There are controls to avoid that.
 
 ## GC Heap Hard Limit
 
-The following configuration knobs will be exposed to enable developers to configure their applications:
+The *GC Heap Hard Limit* is the maximum managed heap size. It only applies when running within a cgroup. By default, it is lower than the cgroup memory constraint (AKA "the cgroup hard limit").
 
-* `GCHeapHardLimit` - specifies a hard limit for the GC heap as an absolute value
-* `GCHeapHardLimitPercent` - specifies a hard limit for the GC heap as a percentage of physical memory that the process is allowed to use
+The following configuration knobs are exposed to configure applications:
+
+* `GCHeapHardLimit` - specifies a hard limit for the GC heap as an absolute value, in bytes (hex value).
+* `GCHeapHardLimitPercent` - specifies a hard limit for the GC heap as a percentage of the cgroup hard limit (hex value).
```

Comment:
Would it make sense to add an example of what a `GCHeapHardLimit` hex value looks like? Does it have a prefix like `0xf00d` or `f00d`?

For `GCHeapHardLimitPercent`, does it range from 0 to 100 (or hex `0x0` to `0x64`)?

Reply:
We have the full documentation with examples and other details at https://docs.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector

Should we move some of these edits to the official docs instead?

Reply:
Added the link to the doc.